### ISSUE TITLE
It seems that iterating over a deterministic dataloader changes the seed state

### ISSUE THREAD

**Issue by LinglanZhao** [pytorch/pytorch#122697]
### 🐛 Describe the bug

By switching **eval_testset = False** to **eval_testset = True** in the codes (i.e., whether iterating over a deterministic dataloader with shuffle=False and fixed transforms), the program with a fixed seed gets different results:
```python
import torch
import numpy as np
import random
import torchvision
import torchvision.transforms as transforms

def set_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=32,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=32,
                                         shuffle=False, num_workers=2)

set_seed(42)
eval_testset = True

for epoch in range(10):
    if epoch >= 2:
        break # for the purpose of saving time

    for i, data in enumerate(trainloader):
        inputs, labels = data
        # train the model ...
        if i == 0:
            print("eval_testset=%s, epoch=%d, iter=%d, loaded labels=%s"%(eval_testset, epoch, i, labels.tolist()))

    if eval_testset:
        with torch.no_grad():
            for data in testloader:
                images, labels = data
                # test the model ...
```
**eval_testset = False:** 
```
eval_testset=False, epoch=0, iter=0, loaded labels=[6, 0, 4, 1, 2, 7, 9, 4, 7, 8, 4, 5, 6, 0, 4, 2, 0, 1, 6, 1, 4, 3, 2, 3, 2, 4, 0, 7, 5, 1, 8, 6]
eval_testset=False, epoch=1, iter=0, loaded labels=[8, 7, 1, 6, 2, 3, 0, 7, 7, 2, 9, 8, 8, 7, 5, 9, 0, 6, 9, 6, 7, 5, 9, 3, 7, 8, 9, 4, 4, 4, 0, 7]
```

**eval_testset = True:**
```
eval_testset=True, epoch=0, iter=0, loaded labels=[6, 0, 4, 1, 2, 7, 9, 4, 7, 8, 4, 5, 6, 0, 4, 2, 0, 1, 6, 1, 4, 3, 2, 3, 2, 4, 0, 7, 5, 1, 8, 6]
eval_testset=True, epoch=1, iter=0, loaded labels=[9, 5, 7, 2, 2, 3, 3, 6, 2, 4, 2, 4, 1, 4, 1, 1, 6, 9, 7, 6, 9, 4, 8, 1, 3, 4, 7, 9, 3, 9, 3, 9]
```
I am wondering how can a deterministic dataloader change the seed state of the codes (i.e., resulting in different outputs).

### Versions

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0            py39h5eee18b_1  
[conda] mkl_fft                   1.3.8            py39h5eee18b_0  
[conda] mkl_random                1.2.4            py39hdb19cb5_0  
[conda] numpy                     1.26.3           py39h5f9d8c6_0  
[conda] numpy-base                1.26.3           py39hb5e798b_0  
[conda] pytorch                   1.13.1          py3.9_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.13.1               py39_cu117    pytorch
[conda] torchvision               0.14.1               py39_cu117    pytorch

cc @andrewkho @gokulavasan @SsnL @VitalyFedyunin @dzhulgakov @mruberry @kurtamohler @ejguan

**Comment by LinglanZhao** [comment]
[Discussions](https://discuss.pytorch.org/t/why-iterating-over-a-deterministic-dataloader-changes-the-seed-state/199610/2) on PyTorch Forums seem to explain the behavior.

**Comment by xiaosuyu1997** [comment]
I have also noticed this issue and narrow it to a more clear case as follows
```
torch.manual_seed(0)
print(torch.randint(0, 200, (1, )).item())
# 44

torch.manual_seed(0)
data = torch.arange(100)
dl = DataLoader(data, batch_size=4, shuffle=False)

iter(dl)
print(torch.randint(0, 200, (1, )).item())
# 133
```
And the line taking part may be [here](https://github.com/pytorch/pytorch/blob/a4b02a15d33cdd946d95af51f48bb0640a6abcb5/torch/utils/data/dataloader.py#L605), as discussed in [discuss](https://discuss.pytorch.org/t/why-iterating-over-a-deterministic-dataloader-changes-the-seed-state/199610/6)

**Comment by xiaosuyu1997** [comment]
Reason for this behavios is already discussed in docs: found [here](https://pytorch.org/docs/stable/data.html#data-loading-randomness)

![image](https://github.com/pytorch/pytorch/assets/33937512/f225462c-d008-4f21-9c52-52b1b860d9e5)

**Comment by gokulavasan** [comment]
@xiaosuyu1997 To me this definitely is a bug that should be fixed. But it will impact (or rather won't impact) the seed state and might break things for users who might be relying on this buggy behavior. Cc @albanD @andrewkho on any inputs on this. 

@xiaosuyu1997 Would you be interested in working on the fix?

**Comment by andrewkho** [comment]
@gokulavasan I'm not totally convinced that this needs to be fixed, users can simply pass a generator to the shuffle=True dataloader to get the desired behaviour. Although _base_seed is only used for multiprocess iterators, IMO the behaviour should still be consistent between num_workers=0 and num_workers>0.

**Comment by kurtamohler** [comment]
I might be misunderstanding something, but it seems that `DataLoader` is expected to change the global RNG seed by default, so it is not a bug. The [reproducibility page](https://pytorch.org/docs/stable/notes/randomness.html#dataloader) in the docs shows an example of how to make `DataLoader` avoid changing the global seed. As others have pointed out, the main thing is to provide a separate generator to each `DataLoader`.

Here is a modified version of the script in the issue description which runs the training loop twice--once with `eval_testset=False`, and once with `eval_testset=True` (just to avoid the manual step of changing it by hand). I've applied the solution that the reproducibility doc recommends.

<details><summary>Click to expand/collapse</summary>

```python
import torch
import numpy as np
import random
import torchvision
import torchvision.transforms as transforms

torch.use_deterministic_algorithms(True)

def set_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)

def seed_worker(worker_id):
    worker_seed = torch.initial_seed() % 2**32
    set_seed(worker_seed)

for eval_testset in [False, True]:
    print('-------------------')
    print(f'eval_testset: {eval_testset}')
    print('-------------------')
    transform = transforms.Compose(
        [transforms.ToTensor(),
         transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

    train_g = torch.Generator()
    test_g = torch.Generator()
    train_g.manual_seed(102)
    test_g.manual_seed(234)

    trainset = torchvision.datasets.CIFAR10(
        root='./data',
        train=True,
        download=True,
        transform=transform)

    trainloader = torch.utils.data.DataLoader(
        trainset,
        batch_size=32,
        shuffle=True,
        num_workers=2,
        worker_init_fn=seed_worker,
        generator=train_g)

    testset = torchvision.datasets.CIFAR10(
        root='./data',
        train=False,
        download=True,
        transform=transform)

    testloader = torch.utils.data.DataLoader(
        testset,
        batch_size=32,
        shuffle=False,
        num_workers=2,
        worker_init_fn=seed_worker,
        generator=test_g)

    set_seed(42)

    for epoch in range(10):
        if epoch >= 2:
            break # for the purpose of saving time

        for i, data in enumerate(trainloader):
            inputs, labels = data
            # train the model ...
            if i == 0:
                print("eval_testset=%s, epoch=%d, iter=%d, loaded labels=%s"%(eval_testset, epoch, i, labels.tolist()))

        if eval_testset:
            with torch.no_grad():
                for data in testloader:
                    images, labels = data
                    # test the model ...
```

</details>

I get the following output, showing that it now produces the same results in both cases:

```
-------------------
eval_testset: False
-------------------
eval_testset=False, epoch=0, iter=0, loaded labels=[5, 4, 8, 6, 5, 9, 7, 2, 4, 0, 1, 1, 1, 1, 4, 6, 2, 6, 5, 3, 0, 0, 2, 6, 8, 4, 6, 6, 7, 5, 8, 9]
eval_testset=False, epoch=1, iter=0, loaded labels=[7, 4, 5, 7, 2, 3, 4, 0, 5, 9, 6, 1, 9, 1, 9, 0, 2, 2, 0, 8, 0, 8, 1, 8, 8, 4, 3, 1, 5, 7, 2, 0]
-------------------
eval_testset: True
-------------------
eval_testset=True, epoch=0, iter=0, loaded labels=[5, 4, 8, 6, 5, 9, 7, 2, 4, 0, 1, 1, 1, 1, 4, 6, 2, 6, 5, 3, 0, 0, 2, 6, 8, 4, 6, 6, 7, 5, 8, 9]
eval_testset=True, epoch=1, iter=0, loaded labels=[7, 4, 5, 7, 2, 3, 4, 0, 5, 9, 6, 1, 9, 1, 9, 0, 2, 2, 0, 8, 0, 8, 1, 8, 8, 4, 3, 1, 5, 7, 2, 0]
```

Also, here is a modified version of this simpler script from [this comment](https://github.com/pytorch/pytorch/issues/122697#issuecomment-2049979534), also with the recommended changes:

<details><summary>Click to expand/collapse</summary>

```python
import torch

torch.manual_seed(0)
print(torch.randint(0, 200, (1, )).item())

torch.manual_seed(0)
data = torch.arange(100)

dl = torch.utils.data.DataLoader(
    data, batch_size=4, shuffle=False, generator=torch.Generator())

iter(dl)
print(torch.randint(0, 200, (1, )).item())
```

</details>

It produces the following output, again showing that it now produces the same result in both cases:

```
44
44
```

**Comment by kurtamohler** [comment]
Actually, I realize now that the real issue is that it seems like `DataLoader(..., shuffle=False)` should not need to generate any random numbers at all. If it really is not needed, then we should fix that, and I think I know of a way to do it--just create a temporary generator to use when initializing the `DataLoader._base_seed`. This temporary generator should be initialized with the current generator's `initial_seed` in order to be deterministic. I'll go ahead and submit a PR that does this

**Comment by andrewkho** [comment]
@kurtamohler I am still not totally convinced, since eg the dataset or custom samplers may also consume RNG states. Consider a contrived example where the dataloader always uses the current global RNG state to initialize and Multiprocessing is on, all workers will start with the same initial seed, it would be the equivalent of the below, and would give different behaviour to the case where num_workers = 0, which also doesn't seem great

```
class Noise(torch.utils.data.Dataset):
  def __getitem__(self, i):
    return i + torch.normal(torch.tensor([0.]), 0.1).item()
  def __len__(self):
    return 10

for num_workers in (0, 2):
  print(f"{num_workers=}")
  g = torch.Generator()
  g.set_state(torch.get_rng_state())
  dl1 = DataLoader(Noise(), batch_size=10, shuffle=False, generator=g, num_workers=num_workers)
  print("  ", next(iter(dl1)).tolist())
  g.set_state(torch.get_rng_state())
  dl2 = DataLoader(Noise(), batch_size=10, shuffle=False, generator=g, num_workers=num_workers)
  print("  ", next(iter(dl2)).tolist())
```
prints:
```
num_workers=0
   [-0.09041650593280792, 0.8216783404350281, 1.812673419713974, 2.952524356544018, 3.846688210964203, 4.938999719917774, 5.827712580561638, 7.025675931945443, 7.959148418158293, 9.050459019839764]
   [0.006950178649276495, 1.0894662737846375, 1.9957839655689895, 2.9510386176407337, 4.032881621271372, 4.951441291719675, 5.950072541832924, 7.041502069681883, 8.040296126157045, 8.934742152690887]
num_workers=2
   [-0.1088351383805275, 1.1863674074411392, 1.9026493057608604, 3.100455179810524, 3.8256215900182724, 4.943101979792118, 6.0631778836250305, 7.117449291050434, 7.946653675287962, 8.95017347112298]
   [-0.1088351383805275, 1.1863674074411392, 1.9026493057608604, 3.100455179810524, 3.8256215900182724, 4.943101979792118, 6.0631778836250305, 7.117449291050434, 7.946653675287962, 8.95017347112298]
```

**Comment by kurtamohler** [comment]
@andrewkho, that is a good argument, I agree! I'll close my PR

**Comment by andrewkho** [comment]
@kurtamohler thanks! Yes happy to hear more discussion before closing the issue
